filters/awssigner/awssigv4/sigv4.go

```diff
@@ -0,0 +1,193 @@
+	}
+
+	disableSessionTokenStr, ok := args[4].(string)
+	if !ok {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	disableHeaderHoisting, err := strconv.ParseBool(disableHeaderHoistingStr)
+	if err != nil {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	disableURIPathEscaping, err := strconv.ParseBool(disableURIPathEscapingStr)
+	if err != nil {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	disableSessionToken, err := strconv.ParseBool(disableSessionTokenStr)
+	if err != nil {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	return &awsSigV4Filter{
+		region:                 region,
+		service:                service,
+		disableHeaderHoisting:  disableHeaderHoisting,
+		disableURIPathEscaping: disableURIPathEscaping,
+		disableSessionToken:    disableSessionToken,
+	}, nil
+}
+
+/*
+sigV4Filter is a request filter that signs the request.
+In case a non empty is body is present in request,
+the body is read and signed. The body is later reassigned to request. Operators should ensure
+that body size by all requests at any point of time is not more than the memory limit of skipper.
+*/
+func (f *awsSigV4Filter) Request(ctx filters.FilterContext) {
+	req := ctx.Request()
+
+	logger := log.WithContext(req.Context())
+
+	signer := NewSigner()
+
+	accessKey := getAndRemoveHeader(ctx, accessKeyHeader, req)
+	if accessKey == "" {
+		return
+	}
+
+	secretKey := getAndRemoveHeader(ctx, secretHeader, req)
+	if secretKey == "" {
+		return
+	}
+
+	sessionToken := getAndRemoveHeader(ctx, sessionHeader, req)
+	if sessionToken == "" {
+		return
+	}
+
+	timeStr := getAndRemoveHeader(ctx, timeHeader, req)
+	if timeStr == "" {
+		return
+	}
+
+	time, err := time.Parse(time.RFC3339, timeStr) //2012-11-01T22:08:41+00:00
+
+	if err != nil {
+		logger.Log(log.ErrorLevel, "time was not in RFC3339 format")
+		return
+	}
+
+	hashedBody, body, err := hashRequest(ctx, req.Body)
+	if err != nil {
+		logger.Log(log.ErrorLevel, fmt.Sprintf("error occured while hashing the body %s", err.Error()))
+		return
+	}
+	creds := internal.Credentials{
+		AccessKeyID:     accessKey,
+		SessionToken:    sessionToken,
+		SecretAccessKey: secretKey,
+	}
+
+	optfn := func(options *SignerOptions) {
+		options.DisableHeaderHoisting = f.disableHeaderHoisting
+		options.DisableSessionToken = f.disableSessionToken
+		options.DisableURIPathEscaping = f.disableSessionToken
+		options.Ctx = ctx.Request().Context()
+	}
+	//modifies request inplace
+	signer.SignHTTP(creds, req, hashedBody, f.service, f.region, time, optfn)
+
+	ctx.Request().Body = io.NopCloser(body) //ATTN: custom close() and read() set by skipper or previous filters are lost
+}
+
+func (f *awsSigV4Filter) Response(ctx filters.FilterContext) {}
+
+func hashRequest(ctx filters.FilterContext, body io.Reader) (string, io.Reader, error) {
+	h := sha256.New()
+	if body == nil {
+		body = strings.NewReader("{}") // as per specs https://github.com/aws/aws-sdk-go-v2/blob/main/aws/signer/v4/v4.go#L261
```

Comment:
please use links with commit hash such that these are stable.
I also wonder if `body==nil` is the way to go and I don't understand why you set `"{}"`

see also: https://go.dev/play/p/gbdvCM6LSEB

Reply:
I believe this is outdated file, the latest commit has [empty string](https://github.com/zalando/skipper/pull/3070/files#diff-c6d3b0b66b89682953361ca4e51ce39b1ac731a2301e36c5849cf13a0afa975bR166) , but `http.NoBody` is better as it is encapsulated ReadCloser with zero bytes. Will implement, thanks